### PR TITLE
Add PowToExponentiationFixer

### PR DESCRIPTION
**Pull request by SpacePossum**
Expected
```php
<?php 1**2;
```
Input
```php
<?php pow(1,2);
```

### CODE REVIEW

**Comment by gharlan** [comment]
`pow(1 + 2, 3 * 4)` should be `(1 + 2) ** (3 * 4)`

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            for ($i = $argumentStart; $i <= $argumentEnd; ++$i) {
+                if ($tokens[$i]->isWhitespace()
+                    || $tokens[$i]->isComment()
+                    || $tokens[$i]->isGivenKind(array(T_LNUMBER, T_VARIABLE, T_STRING, T_NS_SEPARATOR))
+                ) {
+                    continue;
+                }
+
+                if ($tokens[$i]->equals('(')) {
+                    $i = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $i);
```
Comment (gharlan) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: tokens between `[` and `]` could be skipped, too.
`pow($a[1 + 2], $b[3 + 4])` -> `$a[1 + 2] ** $b[3 + 4]`

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        // clean up the function call tokens prt. I
+        $tokens[$closeParenthesisIndex]->clear();
+
+        // check if the arguments need to be wrapped in parenthesis
+        foreach (array_reverse($arguments, true) as $argumentStart => $argumentEnd) {
+            $needsParenthesis = false;
+            for ($i = $argumentStart; $i <= $argumentEnd; ++$i) {
+                if ($tokens[$i]->isWhitespace()
+                    || $tokens[$i]->isComment()
+                    || $tokens[$i]->isGivenKind(array(T_LNUMBER, T_VARIABLE, T_STRING, T_NS_SEPARATOR))
```
Comment (gharlan) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: I think this list could be much longer:

```
$ // for variable variables
++
--
::
->
T_CONSTANT_ENCAPSED_STRING
T_DNUMBER
CT_NAMESPACE_OPERATOR
T_LINE // haha: pow($x, __LINE__) ;)
// and some more without any real use case (in pow()) ;)
```

Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: great feedback, thanks! I took the whitelist approach as to be future prove for when new tokens are added to PHP, but indeed the list can be made longer :)
will update soon

Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: note sure how `CT_NAMESPACE_OPERATOR` could be used
(btw. T_LINE ? I might skip that one ;) )

**Comment by SpacePossum** [comment]
Expect the test to fail.
I need to find the `pow()`-calls reversed, i.e. last first and iterate back to the first token.
Will finish when I got some time. Comments on anything else are welcome ^-^

File: src/Fixer/Alias/PowToExponentiationFixer.php (hunk 2 above)
Comment (gharlan) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: `CT_NAMESPACE_OPERATOR`:
```
pow(namespace\Foo::BAR, 2);
```

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+     * @param int    $argumentEndIndex
+     *
+     * @return bool
+     */
+    private function isParenthesisNeeded(Tokens $tokens, $argumentStartIndex, $argumentEndIndex)
+    {
+        static $allowedKinds;
+
+        if (null === $allowedKinds) {
+            $allowedKinds = array(T_DOUBLE_CAST, T_INT_CAST, T_CONSTANT_ENCAPSED_STRING, T_INC, T_DEC, T_DNUMBER, T_OBJECT_OPERATOR, T_LNUMBER, T_VARIABLE, T_STRING, T_NS_SEPARATOR, T_WHITESPACE, T_COMMENT, T_DOC_COMMENT, T_DOUBLE_COLON);
```
Comment (keradus) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: it contains only compile-level known data, could be defined inplace (`śtatic $foo = [...]`)

File: src/Fixer/Alias/PowToExponentiationFixer.php (hunk 1 above)
Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: nice catch, done https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2192/files#diff-b412676c923661ef450f4a0903c5442aR140

File: src/Fixer/Alias/PowToExponentiationFixer.php (hunk 2 above)
Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: done, including `T_LINE` ;)

File: src/Fixer/Alias/PowToExponentiationFixer.php (hunk 3 above)
Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: :+1: fixed

**Comment by SpacePossum** [comment]
All done I think.
IMHO fabbot.io is wrong

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            T_COMMENT, T_DOC_COMMENT,
+        );
+
+        for ($i = $argumentStartIndex; $i <= $argumentEndIndex; ++$i) {
+            if ($tokens[$i]->isGivenKind($allowedKinds) || $tokens[$i]->isEmpty()) {
+                continue;
+            }
+
+            if ($tokens[$i]->equals('(')) {
+                $i = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $i);
```
Comment (gharlan) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: why not ignoring all block types by using `$tokens->detectBlockType()`?

`BLOCK_TYPE_ARRAY_SQUARE_BRACE` is missing..

```php
pow([1, 2, 3, 4][$x], 2);
```

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    private function isParenthesisNeeded(Tokens $tokens, $argumentStartIndex, $argumentEndIndex)
+    {
+        static $allowedKinds = array(
+            T_DNUMBER, T_LNUMBER, T_VARIABLE, T_STRING, T_OBJECT_OPERATOR, T_CONSTANT_ENCAPSED_STRING, T_DOUBLE_CAST,
+            T_INT_CAST, T_INC, T_DEC, T_NS_SEPARATOR, T_WHITESPACE, T_DOUBLE_COLON, CT_NAMESPACE_OPERATOR, T_LINE,
+            T_COMMENT, T_DOC_COMMENT,
+        );
+
+        for ($i = $argumentStartIndex; $i <= $argumentEndIndex; ++$i) {
+            if ($tokens[$i]->isGivenKind($allowedKinds) || $tokens[$i]->isEmpty()) {
```
Comment (gharlan) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: missing variable variables (token '$'): `pow($$a, $$b)`

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                }
+            }
+
+            if ($tokens[$i]->isGivenKind(CT_DYNAMIC_PROP_BRACE_OPEN)) {
+                $i = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_DYNAMIC_PROP_BRACE, $i);
+
+                continue;
+            }
+
+            if (($tokens[$i]->equals('-') || $tokens[$i]->equals('+')) && $tokens->getPrevMeaningfulToken($i) < $argumentStartIndex) {
```
Comment (gharlan) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: `equalsAny`

File: tests/Fixer/Alias/PowToExponentiationFixerTest.php [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                '<?php 1**2?>',
+                '<?php pow(1,2)?>',
+            ),
+            array(
+                '<?php 1.2**2.3;',
+                '<?php pow(1.2,2.3);',
+            ),
+            array(
+                '<?php echo -2** 3;',
+                '<?php echo pow(-2, 3);',
```
Comment (gharlan) [inline tests/Fixer/Alias/PowToExponentiationFixerTest.php]: https://3v4l.org/q6ZKt

Comment (SpacePossum) [inline tests/Fixer/Alias/PowToExponentiationFixerTest.php]: thanks for pointing that one out, I find it odd PHP behavior... anyway back to the code

Comment (GrahamCampbell) [inline tests/Fixer/Alias/PowToExponentiationFixerTest.php]: Makes sense that the power operator binds more strongly than the subtraction operator (or the unary equivalent).

Comment (SpacePossum) [inline tests/Fixer/Alias/PowToExponentiationFixerTest.php]: I would reason the subtraction here is not an operation at all, but just the number `-2`

File: src/Fixer/Alias/PowToExponentiationFixer.php (hunk 5 above)
Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: these are in, I'll add another test for those

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 8]
```diff
@@ -198,13 +186,7 @@ private function isParenthesisNeeded(Tokens $tokens, $argumentStartIndex, $argum
                 }
             }
 
-            if ($tokens[$i]->isGivenKind(CT_DYNAMIC_PROP_BRACE_OPEN)) {
-                $i = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_DYNAMIC_PROP_BRACE, $i);
-
-                continue;
-            }
-
-            if (($tokens[$i]->equals('-') || $tokens[$i]->equals('+')) && $tokens->getPrevMeaningfulToken($i) < $argumentStartIndex) {
```
Comment (SpacePossum) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: still missing a test for this case

File: src/Fixer/Alias/PowToExponentiationFixer.php [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    private function findPowCalls(Tokens $tokens)
+    {
+        $candidates = array();
+
+        // Minimal candidate to fix is seven tokens: pow(x,x);
+        $end = count($tokens) - 6;
+
+        // First possible location is after the open token: 1
+        for ($i = 1; $i < $end; ++$i) {
+            if (null === $candidate = $this->find('pow', $tokens, $i, $end)) {
```
Comment (keradus) [inline src/Fixer/Alias/PowToExponentiationFixer.php]: don't assign inside condition

**Comment by keradus** [comment]
Please add the fixer to `PHP70Migration` ruleset

**Comment by SpacePossum** [comment]
>Please add the fixer to PHP70Migration ruleset

This feature was added to PHP 5.6, you sure you want it to a PHP7 modernizing set?

**Comment by keradus** [comment]
Oh! Please make PHP56Migration ruleset then :D

File: tests/AbstractFunctionReferenceFixerTest.php [hunk 10]
```diff
@@ -25,23 +25,23 @@
...
+        $tokens = Tokens::fromCode($code);
         $mock = new AccessibleObject($this->getMockForAbstractClass('\\PhpCsFixer\\AbstractFunctionReferenceFixer'));
 
-        $this->assertSame(
-            $argumentsCount,
-            $mock->countArguments(Tokens::fromCode($code), $openIndex, $closeIndex)
-        );
+        $this->assertSame($argumentsCount, $mock->countArguments($tokens, $openIndex, $closeIndex));
+        $this->assertSame($arguments, $mock->getArguments($tokens, $openIndex, $closeIndex));
+        $this->assertCount($argumentsCount, $arguments);
```
Comment (keradus) [inline tests/AbstractFunctionReferenceFixerTest.php]: this is comparing both inputs. why ?

File: tests/AbstractFunctionReferenceFixerTest.php [hunk 11]
```diff
@@ -25,23 +25,23 @@
     /**
      * @dataProvider provideCases
      */
-    public function testCountArguments($code, $openIndex, $closeIndex, $argumentsCount)
+    public function testCountArguments($code, $openIndex, $closeIndex, $argumentsCount, $arguments)
```
Comment (keradus) [inline tests/AbstractFunctionReferenceFixerTest.php]: now, with `$arguments`, `$argumentsCount` is redundant

**Comment by keradus** [comment]
:+1: except minors

File: tests/AbstractFunctionReferenceFixerTest.php (hunk 11 above)
Comment (SpacePossum) [inline tests/AbstractFunctionReferenceFixerTest.php]: indeed, good catch, removed

**Comment by keradus** [comment]
I let myself push some fixes

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by SpacePossum** [comment]
thanks @keradus : D

**Pull request merged**